python/tritonserver/_api/_tensor.py

```diff
@@ -0,0 +1,594 @@
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import ctypes
+import struct
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Callable, ClassVar, Sequence
+
+import numpy
+from _datautils import (
+    NUMPY_TO_TRITON_DTYPE,
+    TRITON_MEMORY_TYPE_TO_DLPACK_DEVICE_TYPE,
+    TRITON_TO_DLPACK_DTYPE,
+    DLPackObject,
+    _parse_device_or_memory_type,
+)
+from _dlpack import DLDevice, DLDeviceType, DLManagedTensor, c_str_dltensor
+from tritonserver._c.triton_bindings import (
+    InvalidArgumentError,
+    TRITONSERVER_BufferAttributes,
+)
+from tritonserver._c.triton_bindings import TRITONSERVER_DataType as DataType
+from tritonserver._c.triton_bindings import TRITONSERVER_MemoryType as MemoryType
+from tritonserver._c.triton_bindings import UnsupportedError
+
+DeviceOrMemoryType = (
+    tuple[MemoryType, int] | MemoryType | tuple[DLDeviceType, int] | str
+)
+
+try:
+    import cupy
+except ImportError:
+    cupy = None
+
+
+@dataclass
+class MemoryBuffer:
+    """Memory allocated for a Tensor.
+
+    This object does not own the memory but holds a reference to the
+    owner.
+
+    Parameters
+    ----------
+    data_ptr : int
+        Pointer to the allocated memory.
+    memory_type : MemoryType
+        memory type
+    memory_type_id : int
+        memory type id (typically the same as device id)
+    size : int
+        Size of the allocated memory in bytes.
+    owner : Any
+        Object that owns or manages the memory buffer.  Allocated
+        memory must not be freed while a reference to the owner is
+        held.
+
+    Examples
+    --------
+    >>> buffer = MemoryBuffer.from_dlpack(numpy.array([100],dtype=numpy.uint8))
```

Comment:
The examples look great!

Other examples don't have the `>>>`, so removing for consistency
```suggestion
    buffer = MemoryBuffer.from_dlpack(numpy.array([100],dtype=numpy.uint8))
```

Reply:
On the topic of doctest or pydantic/pytest-examples mentioned, I don't know if these `>>>` will impact the "running" of the examples as well

Reply:
will normalize examples after merge if ok